modules/image-loading-optimization/module.php

```diff
@@ -0,0 +1,333 @@
+<?php
+namespace Elementor\Modules\ImageLoadingOptimization;
+
+use Elementor\Core\Base\Module as BaseModule;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly
+}
+
+class Module extends BaseModule {
+	/**
+	 * @var int Minimum square-pixels threshold.
+	 */
+	private $min_priority_img_pixels;
+
+	/**
+	 * @var int The number of content media elements to not lazy-load.
+	 */
+	private $omit_threshold;
+
+	/**
+	 * @var array Keep a track of images for which loading optimization strategy were computed.
+	 */
+	private static $image_visited = [];
+
+	/**
+	 * Get Module name.
+	 */
+	public function get_name() {
+		return 'image-loading-optimization';
+	}
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		parent::__construct();
+
+		$this->min_priority_img_pixels = 50000;
+		$this->omit_threshold = 3;
+
+		// Stop wp core logic.
+		add_action( 'init', [ $this, 'stop_core_fetchpriority_high_logic' ] );
+		add_filter( 'wp_lazy_loading_enabled', '__return_false' );
+
+		// Run optimization logic on header.
+		add_action( 'get_header', [ $this, 'set_buffer' ] );
+		// Ensure buffer is flushed (if any) before the content logic.
+		add_filter( 'the_content', [ $this, 'flush_header_buffer' ], 0 );
```

Comment:
Yeah, it's also important to make sure that the flushing doesn't accidentally flush another output buffer that another plugin may have opened (e.g. a caching plugin). I suppose that probably won't be the case since `get_header` is much later than caching plugins start their output buffering. But another plugin could start an output buffer at the `wp_head` action, in which case the call to `ob_end_flush()` won't flush the right buffer: it would flush the outermost buffer(s) not the one that was started at `get_header`. To do that, there would probably need to be a loop that checks `ob_get_status()` until it gets to `handle_buffer_content` as @joemcgill suggests.

Nevertheless, I wonder if this will be free of all side effects, as again it would cause buffers to get flushed perhaps prematurely for some plugins. 

What if instead of buffering just the header and just the footer, it instead buffers the whole page? Then it could inject some HTML comment markers for where `the_content` first starts and last ends, and then the output buffer processing can just process everything up to the first comment (the header) and everything after the last (the footer)?

Reply:
@joemcgill Thanks for pointing this out. I have added the check [here](https://github.com/kt-12/elementor/pull/1/files#diff-b35207b5df0f5802570d77f8eb0f6b5d697b911ba1049355b6ff34a42b55de24R87). 
@westonruter you have a valid point, to test this I did previously set `ob_start` in `init` and let PHP end script take the default course and call the handler. 
It did work fine but I didn't go that path for couple of reason (just my theory need to validate it through lab test).

1) print  would only happen towards the end, so I wanted to see if it effects TTFB of a site as compared to the same site printing out html in parts. I did read from some stackoverflow(not so authentic claim) that they had issues of page loading slow due to full page buffering they implemented.
2) Need to check effects of full page buffering on memory -> considering the fact that there can be sites that run on low-end servers like `aws lightsail $3.50` or `$5` servers. 
I feel full page buffer must go in combination with a good caching logic.
3) I wasn't sure what happens to a large page with a lower buffer size - will the the flush happen in between and handle gets buffer in chunks instead of a single buffer. 
4) Also we could avoid a loop when we use `wp_content_img_tag` as compared to buffer - content would have most number of images for majority of sites.

Reply:
Thanks for the update. Ideally, think Weston's right, that it would be best to process the whole page in a single buffer, which was our original objective. Though I agree with you that finding an approach that reliably captures all of the page output, including any custom theme logic that happens prior to Elementor's output filtering, and does not interfere with any output buffering done by third parties, is a tricky problem to solve. Let's keep thinking about how to make the approach more robust, but I don't see that as an immediate blocker for now.